### PR TITLE
*(dm): recycle resources when initialization of checkpoint in syncer, loader, shard-group-keeper failed

### LINKED ISSUE #4353: *(dm): db not closed when init failed
### description

I found the same problem in more places same as https://github.com/pingcap/tiflow/issues/4142

https://github.com/pingcap/tiflow/blob/4a06fe3104fc5ee37b4328e8f6a8c2cbfdd40199/dm/loader/checkpoint.go#L103-L126

https://github.com/pingcap/tiflow/blob/4a06fe3104fc5ee37b4328e8f6a8c2cbfdd40199/dm/syncer/checkpoint.go#L413-L425

https://github.com/pingcap/tiflow/blob/4a06fe3104fc5ee37b4328e8f6a8c2cbfdd40199/dm/syncer/sharding_group.go#L429-L441

didn't close db once err return

### PR DESCRIPTION
**Pull request by niubell**
### What problem does this PR solve?

Issue Number: close #4353 

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
None.
```

### CODE REVIEW

**Review by okJiang — APPROVED** [review]

File: dm/loader/checkpoint.go [hunk 1]
```diff
@@ -101,7 +102,18 @@ type RemoteCheckPoint struct {
...
+	rollbackHolder := fr.NewRollbackHolder("loader")
+	defer func() {
+		if err != nil {
+			rollbackHolder.RollbackReverseOrder()
+		}
+	}()
+
+	db, dbConns, err = createConns(tctx, cfg, cfg.Name, cfg.SourceID, 1)
 	if err != nil {
 		return nil, err
```
Comment (Ehco1996) [inline dm/loader/checkpoint.go:118]: seems if err != nil  we have not add add rollbcak func into`rollbackHolder` yes , so there is no rollbak in defer🤔, can we add a ut to test the rowback func really works?

File: dm/loader/checkpoint.go [hunk 2]
```diff
@@ -116,6 +128,7 @@ func newRemoteCheckPoint(tctx *tcontext.Context, cfg *config.SubTaskConfig, id s
 		logger:         tctx.L().WithFields(zap.String("component", "remote checkpoint")),
 	}
 	cp.restoringFiles.pos = make(map[string]map[string]FilePosSet)
+	rollbackHolder.Add(fr.FuncRollback{Name: "CloseRemoteCheckPoint", Fn: cp.Close})
 
 	err = cp.prepare(tctx)
```
Comment (Ehco1996) [inline dm/loader/checkpoint.go:133]: how about close DB in func `cp.prepre`, so we need't and rollback func anymore

Comment (niubell) [reply dm/loader/checkpoint.go:133]: A uniform rollback can handle all problems like this in the future, i had done it with close in prepare, but i think rollback is better.

File: dm/loader/checkpoint.go (hunk 1 above)
Comment (niubell) [reply dm/loader/checkpoint.go:118]: > ~seems if err != nil we have not add add rollbcak func into`rollbackHolder` yes , so there is no rollbak in defer🤔, can we add a ut to test the rowback func really works?~
> 
> oh i see it's trying to close db when cp.prepare meet error

bingo

**Comment by niubell** [comment]
I have always tried to sign CLA, but always in vain because of unauthorized email address, i will submit another pull request with a new email address.

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Review by lance6716 — APPROVED** [review]

**Pull request merged**